### PR TITLE
Add HtoE and normalization

### PR DESCRIPTION
**Pull request by ruohai0925**
1. Add H_maxwell = B/mu0 - M in the MacroscopicEvolveE.cpp;

2. Add M normalization in the MacroscopicEvolveM.cpp;

3. Some other small modifications;

### CODE REVIEW

File: Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceSolver.H [hunk 1]
```diff
@@ -77,7 +77,10 @@ class FiniteDifferenceSolver
 
         void MacroscopicEvolveE ( std::array< std::unique_ptr<amrex::MultiFab>, 3>& Efield,
                             std::array< std::unique_ptr<amrex::MultiFab>, 3> const& Bfield,
-                            std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& Jfield,
+#ifdef WARPX_MAG_LLG
+                            std::array< std::unique_ptr<amrex::MultiFab>, 3> const& Mfield,
+#endif
+                std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& Jfield,
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceSolver.H]: fix indentation

File: Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceSolver.H [hunk 2]
```diff
@@ -192,7 +195,10 @@ class FiniteDifferenceSolver
         void MacroscopicEvolveECartesian (
             std::array< std::unique_ptr< amrex::MultiFab>, 3>& Efield,
             std::array< std::unique_ptr< amrex::MultiFab>, 3> const &Bfield,
-            std::array< std::unique_ptr< amrex::MultiFab>, 3> const& Jfield,
+#ifdef WARPX_MAG_LLG
+            std::array< std::unique_ptr< amrex::MultiFab>, 3> const &Mfield,
+#endif
+        std::array< std::unique_ptr< amrex::MultiFab>, 3> const& Jfield,
             amrex::Real const dt,
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceSolver.H:202]: fix indentation

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp [hunk 3]
```diff
@@ -62,6 +62,7 @@ void FiniteDifferenceSolver::MacroscopicEvolveM (
         // static constexpr amrex::Real Ms = 1e4;
         // Real constexpr cons1 = -mag_gamma_interp; // should be mu0*gamma, mu0 is absorbed by B used in this case
         // Real constexpr cons2 = -cons1*alpha/Ms; // factor of the second term in scalar LLG
+        Real M_normalized_error = 0.1; // normalization error of M field for checking
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp]: Can you write some comments on how this value is chosen? if it is used just for debugging, then can you include this in #ifdef AMREX_DEBUG as well? Could this be a user-defined value, which could over-ride a default value if the user wants.

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp [hunk 4]
```diff
@@ -118,7 +119,7 @@ void FiniteDifferenceSolver::MacroscopicEvolveM (
 
               // magnetic material properties mag_alpha and mag_Ms are defined at cell nodes
               // keep the interpolation
-              Real mag_gamma_interp = MacroscopicProperties::macro_avg_to_face(i,j,k,amrex::IntVect(1,0,0),mag_gamma_arr); 
+              Real mag_gamma_interp = MacroscopicProperties::macro_avg_to_face(i,j,k,amrex::IntVect(1,0,0),mag_gamma_arr);
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:122]:
```suggestion
              amrex::Real mag_gamma_interp = MacroscopicProperties::macro_avg_to_face(i,j,k,amrex::IntVect(1,0,0),mag_gamma_arr);
```

Comment (RevathiJambunathan) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:122]: All "Real" are defined in amrex namespace.
We are adopting an explicit definition in all new contributions, where, we add the amrex:: opertor to clarify this.

Please can you add amrex:: to all the Real that is defined in this PR. (no need to clean everywhere.)

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp [hunk 5]
```diff
@@ -138,7 +139,23 @@ void FiniteDifferenceSolver::MacroscopicEvolveM (
               M_xface(i, j, k, 2) += dt * (-PhysConst::mu0 * mag_gamma_interp) * ( M_xface(i, j, k, 0) * Hy_eff - M_xface(i, j, k, 1) * Hx_eff)
                 + dt * Gil_damp * ( M_xface(i, j, k, 0) * ( M_xface(i, j, k, 2) * Hx_eff - M_xface(i, j, k, 0) * Hz_eff)
                 - M_xface(i, j, k, 1) * ( M_xface(i, j, k, 1) * Hz_eff - M_xface(i, j, k, 2) * Hy_eff));
- 
+
+              // temporary normalized magnitude of M_xface field at the fixed point
+              // re-investigate the way we do Ms interp, in case we encounter the case where Ms changes across two adjacent cells that you are doing interp
+              Real M_normalized = sqrt( pow(M_xface(i, j, k, 0),2.0) + pow(M_xface(i, j, k, 1),2.0) + pow(M_xface(i, j, k, 2),2.0) ) /
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp]:
```suggestion
              amrex::Real M_normalized = std::sqrt( std::pow(M_xface(i, j, k, 0),2.0_rt) + std::pow(M_xface(i, j, k, 1),2.0_rt) + std::pow(M_xface(i, j, k, 2),2.0_rt) ) /
```

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp (hunk 3 above)
Comment (RevathiJambunathan) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp]: is "M" for magnetization. if so, can you rename this "mag_normalized_error" to be consistent with other magnetization variables.

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp (hunk 5 above)
Comment (RevathiJambunathan) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp]: same here. "m_normalized" to mag_normalized

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp [hunk 6]
```diff
@@ -62,6 +62,7 @@ void FiniteDifferenceSolver::MacroscopicEvolveM (
         // static constexpr amrex::Real Ms = 1e4;
         // Real constexpr cons1 = -mag_gamma_interp; // should be mu0*gamma, mu0 is absorbed by B used in this case
         // Real constexpr cons2 = -cons1*alpha/Ms; // factor of the second term in scalar LLG
+        Real M_normalized_error = 0.1; // normalization error of M field for checking
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:66]: note... in a future PR let's put this in the MaterialProperties class so we can read in from the inputs file (you can close this comment once you read)

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp [hunk 7]
```diff
@@ -138,7 +139,23 @@ void FiniteDifferenceSolver::MacroscopicEvolveM (
...
                 - M_xface(i, j, k, 1) * ( M_xface(i, j, k, 1) * Hz_eff - M_xface(i, j, k, 2) * Hy_eff));
- 
+
+              // temporary normalized magnitude of M_xface field at the fixed point
+              // re-investigate the way we do Ms interp, in case we encounter the case where Ms changes across two adjacent cells that you are doing interp
+              Real M_normalized = sqrt( pow(M_xface(i, j, k, 0),2.0) + pow(M_xface(i, j, k, 1),2.0) + pow(M_xface(i, j, k, 2),2.0) ) /
+                                  MacroscopicProperties::macro_avg_to_face(i,j,k,amrex::IntVect(1,0,0),mag_Ms_arr);
+
+#ifdef AMREX_DEBUG
+              // check the normalized error
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:148]: Check with Jackie - do we want this to be a DEBUG feature only or always on?
I say always make this on - if people don't want this feature they can set M_normalized_error to a huge value.

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp [hunk 8]
```diff
@@ -138,7 +139,23 @@ void FiniteDifferenceSolver::MacroscopicEvolveM (
...
+              if ( std::abs(1.-M_normalized) > M_normalized_error ){
+                  printf("i = %d, j=%d, k=%d\n", i, j, k);
+                  amrex::Abort("Exceed the normalized error of the M_xface field");
+              }
+#endif
+              // normalize the M_xface field
+              M_xface(i,j,k,0) = M_xface(i,j,k,0)/M_normalized;
+              M_xface(i,j,k,1) = M_xface(i,j,k,1)/M_normalized;
+              M_xface(i,j,k,2) = M_xface(i,j,k,2)/M_normalized;
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp]: change these to "/=" to give the compiler optimization hints?  (same for M_yface and M_zface)

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp [hunk 9]
```diff
@@ -138,7 +139,23 @@ void FiniteDifferenceSolver::MacroscopicEvolveM (
...
+              // temporary normalized magnitude of M_xface field at the fixed point
+              // re-investigate the way we do Ms interp, in case we encounter the case where Ms changes across two adjacent cells that you are doing interp
+              Real M_normalized = sqrt( pow(M_xface(i, j, k, 0),2.0) + pow(M_xface(i, j, k, 1),2.0) + pow(M_xface(i, j, k, 2),2.0) ) /
+                                  MacroscopicProperties::macro_avg_to_face(i,j,k,amrex::IntVect(1,0,0),mag_Ms_arr);
+
+#ifdef AMREX_DEBUG
+              // check the normalized error
+              if ( std::abs(1.-M_normalized) > M_normalized_error ){
+                  printf("i = %d, j=%d, k=%d\n", i, j, k);
+                  amrex::Abort("Exceed the normalized error of the M_xface field");
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:152]: can you also print out the value of M_normalized and M_normalized_error? (same for the instances below)

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp (hunk 6 above)
Comment (ruohai0925) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:66]: Got it!

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveE.cpp [hunk 10]
```diff
@@ -145,9 +173,17 @@ void FiniteDifferenceSolver::MacroscopicEvolveECartesian (
                                                 i, j, k, amrex::IntVect(1,0,0) );
                 amrex::Real mu = T_MacroAlgo::macro_avg_to_edge(i, j, k, amrex::IntVect(2,1,1),
                                                     mu_arr);
-                Ex(i, j, k) = alpha * Ex(i, j, k) + (beta/mu)
+#ifdef WARPX_MAG_LLG
+                Ex(i, j, k) = alpha * Ex(i, j, k) + beta
+                     * ((- T_Algo::DownwardDz(By, coefs_z, n_coefs_z, i, j, k, 0)
+                         + T_Algo::DownwardDy(Bz, coefs_y, n_coefs_y, i, j, k, 0))/PhysConst::mu0
+                 - T_Algo::DownwardDz(M_yface, coefs_z, n_coefs_z, i, j, k, 1)
+                         + T_Algo::DownwardDy(M_zface, coefs_y, n_coefs_y, i, j, k, 2));
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveE.cpp]: indentation

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveE.cpp [hunk 11]
```diff
@@ -157,9 +193,17 @@ void FiniteDifferenceSolver::MacroscopicEvolveECartesian (
                                                 i, j, k, amrex::IntVect(0,1,0) );
                 amrex::Real mu = T_MacroAlgo::macro_avg_to_edge(i, j, k, amrex::IntVect(1,2,1),
                                                     mu_arr);
-                Ey(i, j, k) = alpha * Ey(i, j, k) + (beta/mu)
+#ifdef WARPX_MAG_LLG
+        Ey(i, j, k) = alpha * Ey(i, j, k) + beta
+                     * ((- T_Algo::DownwardDx(Bz, coefs_x, n_coefs_x, i, j, k, 0)
+                         + T_Algo::DownwardDz(Bx, coefs_z, n_coefs_z, i, j, k, 0))/PhysConst::mu0
+                     - T_Algo::DownwardDx(M_zface, coefs_x, n_coefs_x, i, j, k, 2)
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveE.cpp]: indentation

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveE.cpp [hunk 12]
```diff
@@ -145,9 +173,17 @@ void FiniteDifferenceSolver::MacroscopicEvolveECartesian (
...
                 amrex::Real mu = T_MacroAlgo::macro_avg_to_edge(i, j, k, amrex::IntVect(2,1,1),
                                                     mu_arr);
-                Ex(i, j, k) = alpha * Ex(i, j, k) + (beta/mu)
+#ifdef WARPX_MAG_LLG
+                Ex(i, j, k) = alpha * Ex(i, j, k) + beta
+                     * ((- T_Algo::DownwardDz(By, coefs_z, n_coefs_z, i, j, k, 0)
+                         + T_Algo::DownwardDy(Bz, coefs_y, n_coefs_y, i, j, k, 0))/PhysConst::mu0
+                 - T_Algo::DownwardDz(M_yface, coefs_z, n_coefs_z, i, j, k, 1)
+                         + T_Algo::DownwardDy(M_zface, coefs_y, n_coefs_y, i, j, k, 2));
+#else
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveE.cpp:182]: overarching comment: if H_maxwell = B/mu_0 - M, then shouldn't the signs on all the M_ terms be reversed?

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp (hunk 8 above)
Comment (RevathiJambunathan) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp]: alternatively, if you define, amrex::Real inv_mag_normalized = 1.0_rt/M_normalized, then you can do
M_xface(i,j,k,0) *= inv_mag_normalized

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp [hunk 13]
```diff
@@ -179,7 +196,23 @@ void FiniteDifferenceSolver::MacroscopicEvolveM (
               M_yface(i, j, k, 2) += dt * (-PhysConst::mu0 * mag_gamma_interp) * ( M_yface(i, j, k, 0) * Hy_eff - M_yface(i, j, k, 1) * Hx_eff)
                 + dt * Gil_damp * ( M_yface(i, j, k, 0) * ( M_yface(i, j, k, 2) * Hx_eff - M_yface(i, j, k, 0) * Hz_eff)
                 - M_yface(i, j, k, 1) * ( M_yface(i, j, k, 1) * Hz_eff - M_yface(i, j, k, 2) * Hy_eff));
- 
+
+              // temporary normalized magnitude of M_yface field at the fixed point
+              // re-investigate the way we do Ms interp, in case we encounter the case where Ms changes across two adjacent cells that you are doing interp
+              Real M_normalized = sqrt( pow(M_yface(i, j, k, 0),2.0) + pow(M_yface(i, j, k, 1),2.0) + pow(M_yface(i, j, k, 2),2.0) ) /
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp]: Same comment about amrex::  and std:: operators

**Review by RevathiJambunathan — CHANGES_REQUESTED** [review]
Thank you @ruohai0925 for this PR! 
Thank you @ajnonaka for identifying switching signs for Mfield, to account for B/mu_o - M! 
Additionally, I have a few minor suggestions.

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp (hunk 8 above)
Comment (ruohai0925) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp]: Thanks for the suggestions. I will use /= here.

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveE.cpp (hunk 12 above)
Comment (ruohai0925) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveE.cpp:182]: Done!

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp (hunk 13 above)
Comment (ruohai0925) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp]: Sure!

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp (hunk 4 above)
Comment (ruohai0925) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:122]: Ok, I will modify the Real defined in this PR.

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp (hunk 3 above)
Comment (ruohai0925) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp]: Done!

**Comment by ruohai0925** [comment]
All comments in this PR are modified.

**Pull request closed without merging**